### ISSUE TITLE
Null safety feedback: how to learn from the py3 hard lesson, and make it much easier to migrate

### ISSUE THREAD

**Issue by noamraph** [dart-lang/sdk#44466]
Hi,

After playing a a bit with null safety, I think that the current migration plan will cause the migration to null safety to be much more difficult than it could be. I'm copying what I wrote on [reddit](https://www.reddit.com/r/dartlang/comments/kc89h8) with some additions:

The move from python 2 to python 3 was horrible, and I think we can learn from it how to make a large change and how not to, and it seems to me that Dart's current null-safety plans are more like the "how not to" side.

With the current beta, once I upgrade the Dart SDK requirement, all the code in my package becomes broken. This makes upgrading a large project a daunting task, since it all has to be done at once.

I think that the migration to null-safety can be done using deprecation warnings. This means that in the next SDK version, current code will continue to work, but produce deprecation warnings. Those warnings could be solved one by one, until you get code that doesn't produce deprecation warnings. Then, the code without the deprecation warnings would be compatible with the second-next SDK version. Code that produces deprecation warnings could be incompatible with the second-next SDK version.

One thing that the Python community learned the hard way, is that the best way to migrate a large codebase from python 2 to python 3 is by using six, a package that allows you to write code that is compatible with both. I think that that's another good lesson - make to possible to write code that is compatible with both versions of the language.

So, I would suggest something like this:

In Dart 2.12, existing code will continue to work, with `int` meaning a nullable `int?`. You could add a comment (say, `// @nullsafe`) at the beginning of files, so `int` would be non-nullable. It would be possible to use `int?` and `int!` everywhere, so it will be possible to write code that will be compatible both with `// @nullsafe` and without. (This is the equivalent of code using six in Python).

When non-null-safe code puts a nullable value in a non-nullable type, a runtime check will be made, which will cause an exception just like what currently happens when you add an `int` which is null. However, the compiler will issue a deprecation warning, to let you know of the implicit runtime check. With this warning, you will be able to check if the type really should be nullable. If it shouldn't be nullable, you could make the type non-nullable (by replacing `int` with `int!`), which will in turn produce other warnings that you could handle one-by-one. In null-safe code, the compiler would require you to use a non-nullable type.

It would even make sense to show a warning for implicit type declarations in non-null-safe code, so you could either add `?` or `!` to the type to get rid of the warning.

So, the change to null-safety can be this non-breaking change:

* Add the possibility for non-nullable types, by using `int!`.
* Allow to explicitly specify nullable types, by using `int?`.
* Warn on implicit conversions from nullable to non-nullable types
* Warn on implicit nullable types (`int`).

Then, code without warnings will actually be null safe!

In addition, you would be able to declare files as non-nullable by default (for example, by adding `// @nullsafe` at the beginning). This is similar to Python's `from __future__ import` statements. Files with this declaration would behave exactly like the current Dart 2.12. You could add this comment to dart files that don't cause warnings, and they would continue to work exactly the same. There could be a tool that would automatically remove all the unneeded `!` from types afterwards. Actually, I would show a warning on dart files without the comment that don't have any null warnings, suggesting to run the tool. The tool would add the comment and remove the unneeded exclamation marks.

This would mean that instead of needing to migrate your entire code base at once, you would be able to handle one warning at a time, keeping the code base functional all the time. When you took care of all the warnings, you would have null-safe code!

This would also make it much easier to migrate packages with dependencies. With the current Dart 2.12, I can't migrate my package into sound null-safety until all of its dependencies have migrated. This means that it would be much harder to migrate packages. With my suggestion, I could use the dependency package as is, with adding `!` operators when converting a nullable return type to a non-nullable type. This would also make null-safety adoption much faster.

What do you think?

Noam

**Comment by eernstg** [comment]
> once I upgrade the Dart SDK requirement, all the code in my package
> becomes broken. This makes upgrading a large project a daunting task, since
> it all has to be done at once.

If you do not touch the lower bound on the `sdk:` setting in your `pubspec.yaml` there will not be any breakage, so, as you mention, you do have to take action to enable null safety at all.

However, when you enable null safety for your package by updating the `sdk:` lower bound, you can hold back each library of your package by adding a `// @dart = 2.9` comment at the top of the library (several language versions will work, but 2.9 has typically been used so far). This means that said library will stay at the language version 2.9, which doesn't include null safety.

As you make progress and migrate each of the libraries, you'll remove those `// @dart = 2.9` comments. This means that you will end up with a package which is fully migrated, and there are no special markers left in your code. (The alternative would be to add a special comment in each library in order to _enable_ null safety; but that's not so convenient, because you would then have a lot of "enable null safety in this library" markers in your package at the end, and then you'd need to take yet another step in order to make it smoothly and invisibly null safe).

> It would be possible to use `int?` and `int!` everywhere

This approach would of course also be possible, but it would require the migration process to include an intermediate step which would then call for a lot of clean-up work when `@nullsafe` is added. This is because you'd surely want to rely on non-null-by-default types after the migration has completed, in order to avoid having source code where _every single type_ has a marker, either `!` or `?`.

So I think you'll find that the requirement to mark the legacy code as such with `// @dart = 2.9` is more practical: Your package won't need any further adjustments in migrated code.

**Comment by noamraph** [comment]
@eernstg Thank you very much for your thorough reply! After the replies on reddit from @munificent , I understand the situation much better. I think that indeed the migration won't be like the py2-py3 migration, but I think that two changes would be really important in helping people understand that indeed their old packages aren't broken. I'll copy what I wrote in the [comment](https://www.reddit.com/r/dartlang/comments/kc89h8/null_safety_how_to_learn_from_the_py3_hard_lesson/gfslnf4/):

1. Instead of requiring a special comment on legacy files, require a special comment on migrated files. (This is equivalent to Python's `from __future__ import` statements).
2. Instead of requiring the `--no-sound-null-safety` flag if there are legacy files, issue a warning if there are legacy files.

This way, existing working code will continue to work with the new version, without any modifications. This would make everyone realize that the change is indeed not breaking stuff. (Indeed, it wouldn't. With the current plan, it would, even if the fix is not very difficult.)

I think that it's very important to stick to the path of doing updates using deprecation warnings: If a code works without warnings in version N, if means it will do the exact same thing in version N+1, but may produce deprecation warnings. This gives you a "continuous integration" path, where code always works: if you have warnings, fix them (keeping the code working all the time), and then you can be confident to update to the new version.

Now that I understand the soundness issue much better, it seems to me that there is still a path that would allow you to migrate one variable at a time, instead of one file at a time. My understanding is in the new Dart version there are actually three types of types: nullable, non-nullable, and legacy. Let's call them `int?`, `int!` and `int*` respectively. Legacy types are converted automatically to the migrated types. In legacy files, `int` means `int*`, and in migrated files, `int` means `int!`. What if you were allowed to use `int!` and `int?` in legacy files? Then, you could convert legacy variables one at a time. Once there are no more legacy types in a file, Dart could issue a warning that this file could be converted to a migrated file, and a script could adjust the header and remove the unneeded `!` characters.

This would allow you to migrate one variable at a time. Currently, you have to migrate all the variables in a file at once, and if you want to follow all the places a type defined in one file is used, it could mean that you would have to change all the files in your package at once. I think that the path of migrating types one by one, in all the files in which they appear, would allow for much easier migrations, composed of small changes that keep the code working after each change.

To summarize, I think that it's very important that existing code will continue to work in the new version without any changes, and this means the two suggestions I wrote above. I also think that allowing legacy files to use non-legacy types could allow for a significantly easier migration, that won't require large overarching changes that would be a merge nightmare.

**Comment by noamraph** [comment]
Note that even with keeping the principle that code that doesn't cause warnings in version N will continue to work in version N+1, in the second-next Dart version, you could reach a state where you'll need a special comment only for legacy files. In Dart 2.12, show a deprecation warning on files without a "legacy comment", saying that you should add it. Then, in the version after that, you could make the "migrated comment" optional, so a file with no comment will be treated as null-safe.

**Comment by eernstg** [comment]
@noamraph wrote:
> existing working code will continue to work with the new version, without any modifications

This is mainly about granularity. You do have the above property for a complete package: As long as you don't enable null safety by editing your `pubspec.yaml` to have a lower bound `sdk:` of 2.12 or higher, it just keeps working. But if you want to migrate some libraries to null safety then you'll have to edit `pubspec.yaml` and also add something like `// @dart = 2.9` to all the remaining libraries (the ones that won't be migrated right now).

There is no support for migrating a single variable at a time.

However, I suspect that this won't be needed: The typical amount of work in migrating a library using the migration tool isn't that large. I've migrated some tricky packages, but I'd expect a large percentage of cases to be rather straightforward.

If you don't use the migration tool then it may get quite tedious to track down all those data flows that justify adding `?` to many types, so it's better to get "manual control" over the migration by adding `/*?*/` and `/*!*/` hints in the code, which will inform the migration tool that those specific types should be made nullable respectively non-nullable, and then the migration tool will propagate the consequences of that choice to other locations in the code.

So I hope you'll have the same experience that I've had, and I believe many others: It is OK to migrate a whole library at a time using the migration tool. With that, you shouldn't need special support for `int!` in legacy code, etc.

**Comment by noamraph** [comment]
> You do have the above property for a complete package

I see. So I really think it's important to have a script that would add those comments to all the files in a package. Then, make the documentation clear that if you want to use null-safety somewhere in your package, you can change the SDK version and run this script, and everything should work as before.

I think this should be documented as the main migration path: Change the SDK version and add the comment to all the files, commit this refactoring change, and then migrate files one by one using the tool.

(One thing that I find that confuses me, as a newbie, is that "library" means a single file. I understand that it's the same as a "module" in python. Perhaps in the migration document it would help others if the document would explain that library means a single file. It just sounded to me like you need to migrate an entire *package* at once, not a file/library.)

The other breaking change still remains: the need to add `--no-sound-null-safety` as a command-line flag. How about being able to add it to `pubspec.yaml`, and having the compiler issue a warning if it's not needed? Then you would just add it together with changing the SDK version and adding the header comments, and when you no longer have legacy libraries, the compiler would remind you to remove it. (When I tried null safety, I had to find out where to add this argument in android studio, and it wasn't fun).

**Comment by eernstg** [comment]
@leafpetersen, what do you think about adding information to https://dart.dev/null-safety/migration-guide about the use of language version comments like `// @dart = 2.9` and `dart migrate --files=lib/just.dart,lib/some.dart,lib/libraries.dart` to enable migrations where only some libraries are migrated? That page seems to take it for granted that a migration covers a complete package in one step.

@noamraph wrote:
> [it's confusing that] "library" means a single file

Ah, I almost forgot that this isn't the case in all languages. A library is actually one or more files: it includes the library file (which is the one that importers and exporters will refer to) plus possibly a number of `part` files. Those files will be compiled as a single unit, which is also the reason why they have a shared namespace (so code in a part file can depend on code in other part files and in the library file, and vice versa, even private declarations).

I think we should rely on documents like the [Language Tour](https://dart.dev/guides/language/language-tour) to give this kind of information.

> need to add `--no-sound-null-safety` as a command-line flag

That should not be necessary: With a mixed version program `main.dart` (that is, a program that includes both legacy libraries and libraries with null safety), `dart main.dart` must run the program with unsound null checking (it's not supported to execute a mixed program with sound null checking), and it will do so both with and without the option `--no-sound-null-safety`. You can specify `--sound-null-checking` explicitly in order to force execution with sound null checking, and in that case it is an error to try to run a mixed version program.

**Comment by noamraph** [comment]
Regarding migrating in order, from what I understand it seems to me that it should be really fine to migrate not in order, with the only drawback of sometimes having to do extra work after the dependency migrated. This extra work should only involve changing the nullability of some types until the program compiles, and not cause runtime problems.

I think saying that you should wait until all your dependencies have migrated will make null-safety adoption much slower, since large applications will almost certainly have one dependency that won't migrate. If I am correct, I think that it's better to explain the actual cost of migrating before your dependencies, so people could make better decisions and perhaps migrate sooner.

I think I will close this discussion, since I'm investing too much time in it. I want to link to my [discussion](https://www.reddit.com/r/dartlang/comments/kc89h8/null_safety_how_to_learn_from_the_py3_hard_lesson/gfyns0b) with @munificent on reddit, and copy this:

> I think that a bit more patience would paradoxically get you to a sound null-safety world sooner. I think that you have made a tremendous effort to allow backwards compatibility, and small trivial decisions could make the transition much harder, mostly because of the psychology of your actual users.

> I hope I'm wrong. But I think you will realize that many packages remain without null safety because of this impatience, of not wanting null-safe libraries to need an extra line in one SDK version. This will make migrating harder for everyone, and force you to continue supporting null-unsafety longer.

I will write the small changes that I think would do a lot to communicate the actual nature of the change, prevent FUD, and encourage users to migrate:
* Require to add a header line for null-safe libraries in 2.12. Show a deprecation warning saying to add a header line for null-unsafe libraries, since the default would become null-safe in the next version.
* Have a "null soundness" setting in pubspec.yaml, that would default to false in 2.12 with a deprecation warning saying that it should be added, since the default would become true in the next version. Show a compiler warning if it's false and there are no null-unsafe libraries.
* Explain the actual cost of migrating before your dependencies, and make it clear that it's a good choice if your dependencies don't migrate.

Thank you very much for your work, I really enjoy using it to develop an app using Flutter!
Noam

**Comment by parlough** [comment]
Thanks everyone for the discussion here!

At this point, the null safety migration has mostly passed and has been [successful](https://medium.com/dartlang/dart-3-alpha-f1458fb9d232#:~:text=migrate%20existing%20code%20to%20null%20safety) across the Dart ecosystem, so I'm going to close this.

If anyone referencing this hasn't yet migrated to null safety, dart.dev has a guide on [Migrating to null safety](https://dart.dev/null-safety/migration-guide). This page discusses how to migrate by hand or using the migration tool, and has tips to make eventual migration easier. The [Unsound null safety](https://dart.dev/null-safety/unsound-null-safety) page also has some discussion on how to [migrate incrementally](https://dart.dev/null-safety/unsound-null-safety#migrating-incrementally) using language version configuration. Note that this requires Dart 2.19 as Dart 3 and later now only support sound null safety.

Thanks again!
